dot/core/messages.go

```diff
@@ -32,14 +32,21 @@ func (s *Service) validateTransaction(peerID peer.ID, head *types.Header, rt Run
 	externalExt := types.Extrinsic(append([]byte{byte(types.TxnExternal)}, tx...))
 	validity, err = rt.ValidateTransaction(externalExt)
 	if err != nil {
-		if errors.Is(err, runtime.ErrInvalidTransaction) {
-			s.net.ReportPeer(peerset.ReputationChange{
-				Value:  peerset.BadTransactionValue,
-				Reason: peerset.BadTransactionReason,
-			}, peerID)
-		}
-
 		logger.Debugf("failed to validate transaction: %s", err)
+		var txnValidityErr *runtime.TransactionValidityError
+		if errors.As(err, &txnValidityErr) {
+			switch txnValidityErr.Value().(type) {
```

Comment:
nit since we actually don't do anything with the `err` / `txnValidityErr` except checking the underlying error type (`runtime.InvalidTransaction`, `runtime.UnknownTransaction`), should we instead change `rt.ValidateTransaction` to convert the `*runtime.TransactionValidityError` to return an error which can either be `runtime.ErrTransactionInvalid` or `runtime.ErrTransactionUnknown`, wrapping the txnValidityError string i.e. `fmt.Errorf("%w: %s", ErrTransactionInvalid, txnValidityErr.Value())`. We basically push the type switch down the call stack and keep it private in the runtime package, so the Go API is more Go-standard with dumb Go sentinel errors wrapped with context + `fmt.Errorf`. You can then use `errors.Is(err, runtime.ErrTransactionInvalid` in this function here.

Reply:
To make sure that I understand, you are saying to move this logic into the rt.ValidateTransaction function? One thing to consider is that this is just one case, some cases of calling rt.ValidateTransaction responds to if it is a *runtime.TransactionValidityError and not if it's invalid or unknown. In this case, these changes would create more code changes for those cases. 

Part of why I return the whole thing is that it puts the responsibility on the caller, which I think in this case could be nice given the different ways it can be handled. I think I like it as is, but I am flexible if others think this way is better.